Don't keep pointer to QgsPointCloudLayer in QgsPointCloudLayerRenderer and QgsPointCloudLayerProfileGenerator

dvdkon opened the pull request:
## Description

This is a quick change that refactors a small problem in `QgsPointCloudLayerRenderer`. Previously `render()` used a saved `QgsPointCloudLayer` object (and called methods on it and `QgsPointCloudDataProvider`, generating warnings), but that object is not thread safe. Instead we now get everything we need up-front and only keep a refcounted pointer to the thread-safe layer index.

EDIT: I've also made a similar change to `QgsPointCloudLayerProfileGenerator`, since it was pointed out it had the same issue.

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

dvdkon committed:
Don't keep pointer to QgsPointCloudLayer in QgsPointCloudLayerRenderer

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -43,42 +43,47 @@
 QgsPointCloudLayerRenderer::QgsPointCloudLayerRenderer( QgsPointCloudLayer *layer, QgsRenderContext &context )
   : QgsMapLayerRenderer( layer->id(), &context )
-  , mLayer( layer )
+  , mIndex( layer->dataProvider()->index() )
```
let's check whether dataProvider() is not null

dvdkon committed:
Don't use data provider before we check it's non-null

uclaros commented:
Same change needs to be applied to the profile renderer too.
And probably then we can switch from `QGIS_PROTECT_QOBJECT_THREAD_ACCESS_NON_FATAL` to `QGIS_PROTECT_QOBJECT_THREAD_ACCESS`  in `QgsPointCloudLayer::dataProvider()`

uclaros commented on src/core/pointcloud/qgspointcloudlayerrenderer.h:
```diff
@@ -94,7 +95,11 @@ class CORE_EXPORT QgsPointCloudLayerRenderer: public QgsMapLayerRenderer
     QList< QgsMapClippingRegion > mClippingRegions;
+
     const QVector< QgsPointCloudSubIndex > mSubIndexes;
+    std::optional<QgsPointCloudIndex> mOverviewIndex;
```
Just curious, is there a benefit of using `std::optional` instead of just a `QgsPointCloudIndex()` / `QgsPointCloudIndex( nullptr )`, which evaluates as `false`?

dvdkon replied:
Good catch. There isn't, I just personally default to using `std::optional` when a member can be explicitly left out, and don't much like putting `nullptr` in smart pointers.

It *is* redundant (and probably inconsistent with the rest of the codebase), so if you think it's better, I'll change it to the bare index object.

uclaros replied:
That's a nice habit I guess!
It would be ever slightly more readable without it, as `optional` is barely used in the codebase, but I don't have a strong opinion :)

nyalldawson replied:
I'd personally love to see optional used more often -- it's quite an elegant solution to some tricky API considerations... just so long as we don't need to worry about exposure to Python :rofl:

dvdkon commented:
> Same change needs to be applied to the profile renderer too. And probably then we can switch from `QGIS_PROTECT_QOBJECT_THREAD_ACCESS_NON_FATAL` to `QGIS_PROTECT_QOBJECT_THREAD_ACCESS` in `QgsPointCloudLayer::dataProvider()`

Do you mean `QgsPointCloudLayerProfileGenerator`? As far as I can see, `QgsProfilePlotRenderer` doesn't keep any layer object. And in the generator it's more complicated, because it calls `identify()` on the data provider, which is also not thread-safe. So we can't really fix the issue in full by just shuffling objects around like in this PR.

uclaros commented:
> Do you mean `QgsPointCloudLayerProfileGenerator`? As far as I can see, `QgsProfilePlotRenderer` doesn't keep any layer object. And in the generator it's more complicated, because it calls `identify()` on the data provider, which is also not thread-safe. So we can't really fix the issue in full by just shuffling objects around like in this PR.

Yes, I mean `QgsPointCloudLayerProfileGenerator::generateProfile()`, where I believe the main issue lies, as it is called in separate threads for rendering (the comment there pin points the issue as the logic was copied from the 2d renderer).

We can ignore the `identify()` part for now, it is probably called from the main thread.

dvdkon committed:
Don't keep pointer to QgsPointCloudLayer in QgsPointCloudLayerProfileGenerator

dvdkon committed:
Make thread check in QgsPointCloudLayer::dataProvider() fatal

After the last 3 commits there should be no place left that calls this
from a worker thread.

dvdkon commented:
> Yes, I mean `QgsPointCloudLayerProfileGenerator::generateProfile()`, where I believe the main issue lies, as it is called in separate threads for rendering (the comment there pin points the issue as the logic was copied from the 2d renderer).

Thanks, I see it now. I've fixed this method as well and made the `dataProvider()` method fail on non-main-thread access.

uclaros commented:
> Thanks, I see it now. I've fixed this method as well and made the `dataProvider()` method fail on non-main-thread access.

Nice!
You will also need to change `layer->dataProvider()->index()` to the new method `layer->index()`, which returns a different index (editing index) when layer is in editing mode.
Then that  method can also change to the `FATAL` variant!

dvdkon committed:
Use correct index() method to use editing index

dvdkon commented:
> You will also need to change `layer->dataProvider()->index()` to the new method `layer->index()`, which returns a different index (editing index) when layer is in editing mode.

Thanks, I didn't notice that. The test now passes on my machine.

>Then that method can also change to the `FATAL` variant!

Done.

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.h:
```diff
@@ -94,7 +95,11 @@ class CORE_EXPORT QgsPointCloudLayerRenderer: public QgsMapLayerRenderer
+
     const QVector< QgsPointCloudSubIndex > mSubIndexes;
+    std::optional<QgsPointCloudIndex> mOverviewIndex;
+    double mAverageSubIndexWidth;
+    double mAverageSubIndexHeight;
```
```suggestion
    double mAverageSubIndexWidth = 0;
    double mAverageSubIndexHeight = 0;
```

nyalldawson committed:
Initialize members
